### ISSUE TITLE
Textures of size 0 should be allowed

### ISSUE THREAD

**Issue by greggman** [gpuweb/gpuweb#4193]
I'm sure this was already litigated so just point me to the resolution but ...

Lots of apps have a CSS based canvas size. That size can go to zero. Lots of apps set the size of their canvas (its backingstore) to the CSS size.

    canvas.width = canvas.clientWidth  // or equiv using ResizeObserver or getBoundingClientRect

[In Canvas 2D and WebGL this is valid, no problem. In WebGPU it's validation error](https://jsgist.org/?src=fce4c91075bd5a61471a0b381c75f12b). That means there are or will be lots of invalid webgpu apps waiting for the user to size something or add some content that effectively makes a canvas' size go to zero and then, not checking for size = 0 they'll start generating WebGPU errors.

Buffers are allowed to be of size 0, is there a reason textures are not?

I see this:

https://github.com/gpuweb/gpuweb/issues/1200

Maybe `getCurrentTexture` should be special? Maybe it should always return a texture of at least size 1 where as `createTexture` would fail on 0?

Note: Making either of these changes doesn't seem like they'd break any existing content, rather, they'd potentially fix existing broken content.

**Comment by greggman** [comment]
I see this is harder to implement than is sounds. Some APIs do not allow textures of size 0 and there's no easy way to simulate textures of size 0 given `textureDimensions`. 

On the other hand, maybe `getCurrentTexture` returning `max(1, canvas.width)` might not be bad?

In fact, maybe even `max(1, min(device.limits.maxTextureDimension2D, canvas.width)` since the canvas is configured with the device. Why return something that is invalid? As it is, a well written program always has to write this check. You might not think you need it today, but you later expand your app or copy and paste and forget this time bomb is sitting in your code.

**Comment by kainino0x** [comment]
We designed it this way for 2 reasons: first, that it would be easier to do something like this later (because it would be "non-breaking" in that it would only make errors into non-errors), and second that we didn't concretely think that errors would be more problematic than a more complex system where the application either has to duplicate the clamping logic of the browser or it has to query back the actual size it got like in WebGL.

WebGL solves this by clamping at 1 (similar to your proposal), but:
- in practice I don't think applications are using gl.drawingBufferWidth/Height to query the width/height they actually got, and probably just assuming width/height are right
- in WebGPU you have to allocate your own depth/stencil buffer which makes it significantly more common to have extra textures that must match the canvas size - and if you get this wrong then you'll still get a validation error

Also having some validation errors in rendering should hopefully not be catastrophic. The error system in WebGPU means only objects downstream of the canvas texture will actually break. If you resize back to a nonzero size then I _think_ most applications will recover without issue. At least trivial programs certainly do (https://codepen.io/kainino0x/pen/poQNrKb), it's harder to be certain about how people actually write applications in the wild.

**Comment by greggman** [comment]
Maybe to put it another way, it would be nice if there was a path that didn't require the checks.

With this change you could do

```js
canvas.width = canvas.clientWidth;
canvas.height = canvas.clientHeight;

const canvasTexture = context.getCurrentTexture();

const depthTexture = context.createTexture({
  size: [canvasTexture.width, canvasTexture.height],
  ...
```

You'd at least have a path that doesn't require the checks

**Comment by kainino0x** [comment]
Good point you can see what size you actually got by canvasTexture.width/height (we didn't have those yet when we designed this). Maybe I like that proposal, but perhaps with an an explicit option in getCurrentTexture() so you know you're opting into automatic sizing?
